### PR TITLE
feat (api): implement repository layer and RPC provider for live trading

### PR DESCRIPTION
**Pull request by mzkrasner**
# Live Trading Repository Layer and RPC Provider Implementation

## Overview

This PR implements the foundational repository layer and blockchain RPC infrastructure required for live trading functionality as specified in the super minimal implementation spec.

### New Repositories

1. **scan-progress-repository.ts**
   - Methods: getScanProgress, getBatchScanProgress, upsertScanProgress, batchUpsertScanProgress, deleteScanProgressByCompetition
   - Implements efficient batch operations with transactions
   - Uses inArray() for N+1 query prevention

2. **self-funding-alerts-repository.ts**
   - Methods: createSelfFundingAlert, batchCreateSelfFundingAlerts, getAlertsForCompetition, reviewAlert, batchReviewAlerts
   - Includes pagination support for alert retrieval
   - Atomic batch operations using Drizzle transactions

3. **live-competition-config-repository.ts**
   - Methods: createLiveCompetitionConfig, getLiveCompetitionConfig, updateLiveCompetitionConfig, isLiveTradingEnabled
   - Supports batch retrieval for multiple competitions
   - Optional pagination for getAllLiveCompetitionConfigs

### Extended Repositories

- **trade-repository.ts**
  - Added: isTransactionIndexed, createOnChainTrade, batchCreateOnChainTrades, getTradesByType, getOnChainTradesInWindow
  - Maintains existing balance update logic for simulated trades
  - On-chain trades bypass balance updates (tracked separately)

### Alchemy Provider

Implemented a comprehensive Alchemy RPC provider with the following capabilities:

- **Multi-chain Support**: Ethereum, Base, Arbitrum, Optimism, Polygon
- **Core Methods**:
  - getBlockNumber: Retrieve current block for any supported chain
  - getTransactionReceipt: Fetch transaction details
  - getBalance: Get native token (ETH) balance
  - getTokenBalances: Retrieve ERC20 token balances
  - getAssetTransfers: Track incoming/outgoing transfers
  - getTokenDecimals: Dynamically fetch token decimal configuration
  - batchGetTransactionReceipts: Efficient batch transaction queries

### Blockchain Math Utilities

Created `blockchain-math-utils.ts` for handling BigNumber operations:

- Token amount formatting with proper decimal handling
- Gas cost calculations in USD
- Hex to decimal conversions
- Safe arithmetic operations for large numbers
- Decimal caching mechanism for performance

### Environment Variables Added

- `ALCHEMY_API_KEY`: API key for Alchemy RPC access
- `ETH_RPC_URL`, `BASE_RPC_URL`, `ARBITRUM_RPC_URL`, `OPTIMISM_RPC_URL`, `POLYGON_RPC_URL`: Chain-specific RPC endpoints
- `RPC_MAX_RETRIES`, `RPC_RETRY_DELAY_MS`: RPC retry configuration

### Alchemy Provider Tests

Comprehensive test coverage including:
- Basic connectivity tests for all supported chains
- Transaction receipt retrieval
- Token balance queries
- Asset transfer tracking
- Token decimal fetching (USDC, DAI, native ETH)
- Batch operations
- Error handling and edge cases

All 16 tests passing with real Alchemy API integration.

## Documentation Updates

- Updated apps/api/README.md with:
  - Blockchain RPC Provider Support section
  - Alchemy configuration instructions
  - Environment variable reference for RPC settings
  - Technology stack updates
  - Updates to example env files

### CODE REVIEW

File: apps/api/src/services/providers/alchemy.provider.ts [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+  private readonly retryDelayMs: number;
+
+  // Map our chain names to Alchemy Network enum
+  private readonly chainToNetwork: Record<LiveTradingChain, Network> = {
+    eth: Network.ETH_MAINNET,
+    base: Network.BASE_MAINNET,
+    arbitrum: Network.ARB_MAINNET,
+    optimism: Network.OPT_MAINNET,
+    polygon: Network.MATIC_MAINNET,
+    // Note: These chains might not be supported by Alchemy SDK yet
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/alchemy.provider.ts]: resolved in [cf056b6](https://github.com/recallnet/js-recall/pull/1123/commits/cf056b6afb589bb6c5fcd58e019580ede36b8d6c)

File: apps/api/src/services/providers/alchemy.provider.ts [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+    bsc: Network.BNB_MAINNET,
+    avalanche: Network.AVAX_MAINNET,
+    linea: Network.LINEA_MAINNET,
+    zksync: Network.ZKSYNC_MAINNET,
+    scroll: Network.SCROLL_MAINNET,
+    mantle: Network.MANTLE_MAINNET,
+    svm: Network.ETH_MAINNET, // Placeholder - Solana not supported
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/alchemy.provider.ts]: resolved in [cf056b6](https://github.com/recallnet/js-recall/pull/1123/commits/cf056b6afb589bb6c5fcd58e019580ede36b8d6c)

File: apps/api/src/services/providers/alchemy.provider.ts [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+            network,
+            maxRetries: this.maxRetries,
+          });
+          serviceLogger.debug(
+            `[AlchemyProvider] Initialized provider for ${chain}`,
+          );
+        }
+      } catch {
+        serviceLogger.error(
+          `[AlchemyProvider] Failed to initialize ${chain} provider`,
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/alchemy.provider.ts]: resolved in [d5b3f85](https://github.com/recallnet/js-recall/pull/1123/commits/d5b3f85ef7b057f49eaad8d3c76f9579a8a773be)

File: apps/api/src/services/providers/alchemy.provider.ts [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+      // Try to get block number from the first configured chain
+      const testChain = configuredChains[0];
+      if (testChain) {
+        await this.getBlockNumber(testChain);
+        return true;
+      }
+
+      return false;
+    } catch {
+      serviceLogger.error("[AlchemyProvider] Health check failed");
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/alchemy.provider.ts]: resolved in [d5b3f85](https://github.com/recallnet/js-recall/pull/1123/commits/d5b3f85ef7b057f49eaad8d3c76f9579a8a773be)

File: apps/api/src/services/providers/alchemy.provider.ts [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+      });
+
+      // Convert hex result to number
+      if (result && result !== "0x") {
+        return parseInt(result, 16);
+      }
+
+      // Default to 18 if call fails (common for ETH/WETH)
+      return 18;
+    } catch {
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/providers/alchemy.provider.ts]: resolved in [d5b3f85](https://github.com/recallnet/js-recall/pull/1123/commits/d5b3f85ef7b057f49eaad8d3c76f9579a8a773be)

File: apps/api/src/lib/blockchain-math-utils.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  oldValue: string | bigint | number,
+  newValue: string | bigint | number,
+): number {
+  const oldBig = BigInt(oldValue);
+  const newBig = BigInt(newValue);
+
+  if (oldBig === 0n) return newBig > 0n ? 100 : 0;
+
+  const change = ((newBig - oldBig) * 10000n) / oldBig; // Multiply by 10000 for precision
+  return Number(change) / 100; // Divide by 100 to get percentage
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/lib/blockchain-math-utils.ts]: resolved in [d5b3f85](https://github.com/recallnet/js-recall/pull/1123/commits/d5b3f85ef7b057f49eaad8d3c76f9579a8a773be)
